### DISCUSSION TITLE
Clarification of `backendMigrationScript`'s rendering function

### DISCUSSION THREAD

**Discussion by peterbecich** [tathougies/beam#730] [Q&A]
Question about `backendMigrationScript`: https://hackage.haskell.org/package/beam-migrate-0.5.3.1/docs/Database-Beam-Migrate-Simple.html#v:backendMigrationScript

```
-- | Given a function to convert a command to a 'String', produce a script that
-- will execute the given migration. Usually, the function you provide
-- eventually calls 'displaySyntax' to rendere the command.
backendMigrationScript :: BeamSqlBackend be
                       => (BeamSqlBackendSyntax be -> String)
                       -> Migration be a
                       -> String
backendMigrationScript render mig =
  migrateScript ((++"\n") . T.unpack) ((++"\n") . render) (migrationStep "Migration Script" (\() -> mig))
```

Would you provide an example of the `function to convert a command to a 'String'`, the `render` argument?

----------------

I have a test ready to go:
https://github.com/peterbecich/beam/blob/db771f0f58a99d96ac29210fb27da114b537a488/beam-postgres/examples/app/Main.hs#L16

Given:

```
migration :: MigrationSteps Postgres () (CheckedDatabaseSettings Postgres Pagila.Schema.V0002.PagilaDb)
migration = migrationStep "Initial commit" V0001.migration >>>
            migrationStep "Add film actor, inventory, rental table" V0002.migration
```
https://github.com/peterbecich/beam/blob/db771f0f58a99d96ac29210fb27da114b537a488/beam-postgres/examples/src/Pagila/Schema.hs#L18-L20

and a modification to the `backendMigrationScript` function:
```
backendMigrationStepsScript :: BeamSqlBackend be
                            => (BeamSqlBackendSyntax be -> String)
                            -> MigrationSteps be () a
                            -> String
```
https://github.com/peterbecich/beam/blob/db771f0f58a99d96ac29210fb27da114b537a488/beam-migrate/Database/Beam/Migrate/Simple.hs#L342-L350

, I am hoping it will print the SQL generated from these Beam migration `V0001` and `V0002`.

Thanks

**Comment by LaurentRDC** [comment]
Hi Peter,

The key to understanding `(BeamSqlBackendSyntax be -> String)` is to understand `BeamSqlBackendSyntax be`.

`BeamSqlBackendSyntax` is a type family, and also known as a type function, which means that `BeamSqlBackendSyntax be` is a concrete type which depends on the backend. For example, for the [`Postgres` backend, `BeamSqlBackendSyntax be = `PgCommandSyntax`](https://github.com/haskell-beam/beam/blob/249cb07e106dfa8976b9cad96e5f261618855b27/beam-postgres/Database/Beam/Postgres/Types.hs#L172) .

This means that you're looking for a rendered function of the type (`PgCommandSyntax -> String`). 

Drilling deeper, you'll see that any [`PgCommandSyntax` is composed of a `PgSyntax` and some other data that does not need to be printed](https://github.com/haskell-beam/beam/blob/249cb07e106dfa8976b9cad96e5f261618855b27/beam-postgres/Database/Beam/Postgres/Syntax.hs#L231). So, you're looking for a function of type `PgSyntax -> String`.

[Turns out there is something very close called `pgRenderSyntaxScript`](https://github.com/haskell-beam/beam/blob/249cb07e106dfa8976b9cad96e5f261618855b27/beam-postgres/Database/Beam/Postgres/Syntax.hs#L1444), which you can use to build your render function.
